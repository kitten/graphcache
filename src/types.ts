import {
  DocumentNode,
  FragmentDefinitionNode,
  SelectionNode,
  GraphQLOutputType,
  GraphQLWrappingType,
} from 'graphql';

// Helper types
export type NullArray<T> = Array<null | T>;

// GraphQL helper types
export type SelectionSet = ReadonlyArray<SelectionNode>;
export type GraphQLFlatType = Exclude<GraphQLOutputType, GraphQLWrappingType>;
export interface Fragments {
  [fragmentName: string]: void | FragmentDefinitionNode;
}

// Scalar types are not entities as part of response data
export type Primitive = null | number | boolean | string;

export interface ScalarObject {
  __typename?: never;
  [key: string]: any;
}

export type Scalar = Primitive | ScalarObject;

export interface SystemFields {
  __typename: string;
  _id?: string | number | null;
  id?: string | number | null;
}

export type EntityField = undefined | Scalar | Scalar[];
export type DataField = Scalar | Scalar[] | Data | NullArray<Data>;

export interface DataFields {
  [fieldName: string]: DataField;
}

export interface Variables {
  [name: string]: Scalar | Scalar[] | Variables | NullArray<Variables>;
}

export type Data = SystemFields & DataFields;
export type Link<Key = string> = null | Key | NullArray<Key>;
export type ResolvedLink = Link<Data>;
export type Connection = [Variables, string];

export interface FieldInfo {
  fieldKey: string;
  fieldName: string;
  arguments: Variables | null;
}

// This is an input operation
export interface OperationRequest {
  query: DocumentNode;
  variables?: object;
}

export interface ResolveInfo {
  parentTypeName: string;
  parentKey: string;
  parentFieldKey: string;
  fieldName: string;
  fragments: Fragments;
  variables: Variables;
  partial?: boolean;
  optimistic?: boolean;
}

export interface QueryInput {
  query: string | DocumentNode;
  variables?: Variables;
}

export interface Cache {
  /** keyOfEntity() returns the key for an entity or null if it's unkeyable */
  keyOfEntity(data: Data): string | null;

  /** keyOfField() returns the key for a field */
  keyOfField(
    fieldName: string,
    args?: Variables | null | undefined
  ): string | null;

  /** resolve() retrieves the value (or link) of a field on any entity, given a partial/keyable entity or an entity key */
  resolve(
    entity: Data | string | null,
    fieldName: string,
    args?: Variables
  ): DataField;

  /** resolveFieldByKey() returns a field's value on an entity, given that field's key */
  resolveFieldByKey(entity: Data | string | null, fieldKey: string): DataField;

  /** inspectFields() retrieves all known fields for a given entity */
  inspectFields(entity: Data | string | null): FieldInfo[];

  /** invalidateQuery() invalidates all data of a given query */
  invalidateQuery(query: DocumentNode, variables?: Variables): void;

  /** updateQuery() can be used to update the data of a given query using an updater function */
  updateQuery(
    input: QueryInput,
    updater: (data: Data | null) => Data | null
  ): void;

  /** readQuery() retrieves the data for a given query */
  readQuery(input: QueryInput): Data | null;

  /** readFragment() retrieves the data for a given fragment, given a partial/keyable entity or an entity key */
  readFragment(
    fragment: DocumentNode,
    entity: string | Data,
    variables?: Variables
  ): Data | null;

  /** writeFragment() can be used to update the data of a given fragment, given an entity that is supposed to be written using the fragment */
  writeFragment(
    fragment: DocumentNode,
    data: Data,
    variables?: Variables
  ): void;
}

// Cache resolvers are user-defined to overwrite an entity field result
export type Resolver = (
  parent: Data,
  args: Variables,
  cache: Cache,
  info: ResolveInfo
) => DataField | undefined;

export interface ResolverConfig {
  [typeName: string]: {
    [fieldName: string]: Resolver;
  };
}

export type UpdateResolver = (
  result: Data,
  args: Variables,
  cache: Cache,
  info: ResolveInfo
) => void;

export type KeyGenerator = (data: Data) => null | string;

export interface UpdatesConfig {
  Mutation: {
    [fieldName: string]: UpdateResolver;
  };
  Subscription: {
    [fieldName: string]: UpdateResolver;
  };
}

export type OptimisticMutationResolver = (
  vars: Variables,
  cache: Cache,
  info: ResolveInfo
) => null | Data | NullArray<Data>;

export interface OptimisticMutationConfig {
  [mutationFieldName: string]: OptimisticMutationResolver;
}

export interface KeyingConfig {
  [typename: string]: KeyGenerator;
}

export type SerializedEntry = EntityField | Connection[] | Link;

export interface SerializedEntries {
  [key: string]: SerializedEntry;
}

export interface StorageAdapter {
  read(): Promise<SerializedEntries>;
  write(data: SerializedEntries): Promise<void>;
}

export type ErrorCode =
  | 1
  | 2
  | 3
  | 4
  | 5
  | 6
  | 7
  | 8
  | 9
  | 10
  | 11
  | 12
  | 13
  | 15
  | 16
  | 17
  | 18;
