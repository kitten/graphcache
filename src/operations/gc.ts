import { joinKeys } from '../helpers';
import { Store } from '../store';
import { Link } from '../types';

interface Context {
  store: Store;
  visitedLinks: Set<string>;
  visitedRecords: Set<string>;
}

export const gc = (store: Store) => {
  const visitedLinks = new Set<string>();
  const visitedRecords = new Set<string>();
  const ctx = { store, visitedLinks, visitedRecords };

  walkEntity(ctx, 'Query');

  store.records.forEach((_entity, key) => {
    if (!visitedRecords.has(key)) {
      store.remove(key);
    }
  });

  store.links.forEach((_link, key) => {
    if (!visitedLinks.has(key)) {
      store.removeLink(key);
    }
  });
};

const walkEntity = (ctx: Context, key: string) => {
  const { store, visitedRecords, visitedLinks } = ctx;
  const entity = store.find(key);

  if (entity !== null && !visitedRecords.has(key)) {
    visitedRecords.add(key);

    for (const fieldKey in entity) {
      const value = entity[fieldKey];
      if (value === undefined) {
        const childFieldKey = joinKeys(key, fieldKey);
        const link = store.readLink(childFieldKey);
        if (link !== undefined && !visitedLinks.has(childFieldKey)) {
          visitedLinks.add(childFieldKey);
          walkLink(ctx, link);
        }
      }
    }
  }
};

const walkLink = (ctx: Context, link: Link) => {
  if (Array.isArray(link)) {
    link.forEach(childLink => {
      walkLink(ctx, childLink);
    });
  } else if (link !== null) {
    walkEntity(ctx, link);
  }
};
