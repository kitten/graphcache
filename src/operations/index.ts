export { query } from './query';
export { write, writeFragment } from './write';
export { gc } from './gc';
