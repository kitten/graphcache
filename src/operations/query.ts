import warning from 'warning';

import {
  getFragments,
  getMainOperation,
  getSelectionSet,
  normalizeVariables,
  getName,
  getFieldArguments,
  getFieldAlias,
  getOperationName,
} from '../ast';

import {
  Fragments,
  Variables,
  Data,
  DataField,
  Link,
  SelectionSet,
  Completeness,
  OperationRequest,
  NullArray,
} from '../types';

import {
  Store,
  addDependency,
  getCurrentDependencies,
  initStoreState,
  clearStoreState,
} from '../store';

import { SelectionIterator, isScalar } from './shared';
import { joinKeys, keyOfField } from '../helpers';

export interface QueryResult {
  completeness: Completeness;
  dependencies: Set<string>;
  data: null | Data;
}

interface Context {
  result: QueryResult;
  store: Store;
  variables: Variables;
  fragments: Fragments;
}

/** Reads a request entirely from the store */
export const query = (store: Store, request: OperationRequest): QueryResult => {
  initStoreState(0);

  const result = startQuery(store, request);
  clearStoreState();
  return result;
};

export const startQuery = (store: Store, request: OperationRequest) => {
  const operation = getMainOperation(request.query);
  const root: Data = Object.create(null);
  const result: QueryResult = {
    completeness: 'FULL',
    dependencies: getCurrentDependencies(),
    data: root,
  };

  const ctx: Context = {
    variables: normalizeVariables(operation, request.variables),
    fragments: getFragments(request.query),
    result,
    store,
  };

  result.data = readSelection(ctx, 'Query', getSelectionSet(operation), root);

  return result;
};

export const readOperation = (
  store: Store,
  request: OperationRequest,
  data: Data
) => {
  initStoreState(0);

  const operation = getMainOperation(request.query);

  const result: QueryResult = {
    completeness: 'FULL',
    dependencies: getCurrentDependencies(),
    data: null,
  };

  const ctx: Context = {
    variables: normalizeVariables(operation, request.variables),
    fragments: getFragments(request.query),
    result,
    store,
  };

  result.data = readRoot(
    ctx,
    getOperationName(operation),
    getSelectionSet(operation),
    data
  );

  clearStoreState();
  return result;
};

export const readRoot = (
  ctx: Context,
  entityKey: string,
  select: SelectionSet,
  originalData: Data
): Data => {
  if (typeof originalData.__typename !== 'string') {
    return originalData;
  }

  const data = Object.create(null);
  data.__typename = originalData.__typename;

  const iter = new SelectionIterator(entityKey, entityKey, select, ctx);

  let node;
  while ((node = iter.next()) !== undefined) {
    const fieldAlias = getFieldAlias(node);
    const fieldValue = originalData[fieldAlias];

    if (
      node.selectionSet !== undefined &&
      fieldValue !== null &&
      !isScalar(fieldValue)
    ) {
      data[fieldAlias] = readRootField(ctx, getSelectionSet(node), fieldValue);
    } else {
      data[fieldAlias] = fieldValue;
    }
  }

  return data;
};

const readRootField = (
  ctx: Context,
  select: SelectionSet,
  originalData: null | Data | NullArray<Data>
): Data | NullArray<Data> | null => {
  if (Array.isArray(originalData)) {
    const newData = new Array(originalData.length);
    for (let i = 0, l = originalData.length; i < l; i++)
      newData[i] = readRootField(ctx, select, originalData[i]);
    return newData;
  } else if (originalData === null) {
    return null;
  }

  // Write entity to key that falls back to the given parentFieldKey
  const entityKey = ctx.store.keyOfEntity(originalData);
  if (entityKey !== null) {
    const data: Data = Object.create(null);
    return readSelection(ctx, entityKey, select, data);
  } else {
    const typename = originalData.__typename;
    return readRoot(ctx, typename, select, originalData);
  }
};

const readSelection = (
  ctx: Context,
  entityKey: string,
  select: SelectionSet,
  data: Data
): Data | null => {
  const isQuery = entityKey === 'Query';
  if (!isQuery) addDependency(entityKey);

  const { store, variables } = ctx;

  // Get the __typename field for a given entity to check that it exists
  const typename = isQuery ? 'Query' : store.getField(entityKey, '__typename');
  if (typeof typename !== 'string') {
    ctx.result.completeness = 'EMPTY';
    return null;
  }

  data.__typename = typename;

  const iter = new SelectionIterator(typename, entityKey, select, ctx);

  let node;
  while ((node = iter.next()) !== undefined) {
    // Derive the needed data from our node.
    const fieldName = getName(node);
    const fieldArgs = getFieldArguments(node, variables);
    const fieldAlias = getFieldAlias(node);
    const fieldKey = joinKeys(entityKey, keyOfField(fieldName, fieldArgs));
    const fieldValue = store.getRecord(fieldKey);

    if (isQuery) addDependency(fieldKey);

    const resolvers = store.resolvers[typename];
    if (resolvers !== undefined && resolvers.hasOwnProperty(fieldName)) {
      // We have a resolver for this field.
      if (fieldValue !== undefined) {
        data[fieldAlias] = fieldValue;
      }
      const resolverValue = resolvers[fieldName](
        data,
        fieldArgs || {},
        store,
        ctx
      );

      if (node.selectionSet === undefined) {
        // If it doesn't have a selection set we have resolved a property.
        // We assume that a resolver for scalar values implies that this
        // field is always present, so completeness won't be set to EMPTY here
        data[fieldAlias] = resolverValue !== undefined ? resolverValue : null;
      } else {
        // When it has a selection set we are resolving an entity with a
        // subselection. This can either be a list or an object.
        const fieldSelect = getSelectionSet(node);

        data[fieldAlias] = resolveResolverResult(
          ctx,
          resolverValue,
          fieldKey,
          fieldSelect,
          data[fieldAlias] as Data | Data[]
        );
      }
    } else if (node.selectionSet === undefined) {
      // The field is a scalar and can be retrieved directly
      if (fieldValue === undefined) {
        // Cache Incomplete: A missing field means it wasn't cached
        ctx.result.completeness = 'EMPTY';
        data[fieldAlias] = null;
      } else {
        // Not dealing with undefined means it's a cached field
        data[fieldAlias] = fieldValue;
      }
    } else {
      // null values mean that a field might be linked to other entities
      const fieldSelect = getSelectionSet(node);
      const link = store.getLink(fieldKey);

      // Cache Incomplete: A missing link for a field means it's not cached
      if (link === undefined) {
        if (typeof fieldValue === 'object' && fieldValue !== null) {
          // The entity on the field was invalid and can still be recovered
          data[fieldAlias] = fieldValue;
        } else {
          ctx.result.completeness = 'EMPTY';
          data[fieldAlias] = null;
        }
      } else {
        const prevData = data[fieldAlias] as Data;
        data[fieldAlias] = resolveLink(ctx, link, fieldSelect, prevData);
      }
    }
  }

  return data;
};

const resolveResolverResult = (
  ctx: Context,
  result: DataField,
  key: string,
  select: SelectionSet,
  prevData: void | Data | Data[]
) => {
  // When we are dealing with a list we have to call this method again.
  if (Array.isArray(result)) {
    // @ts-ignore: Link cannot be expressed as a recursive type
    return result.map((childResult, index) => {
      const data = prevData !== undefined ? prevData[index] : undefined;
      const indexKey = joinKeys(key, `${index}`);
      return resolveResolverResult(ctx, childResult, indexKey, select, data);
    });
  } else if (result === null) {
    return null;
  } else if (isDataOrKey(result)) {
    // We don't need to read the entity after exiting a resolver
    // we can just go on and read the selection further.
    const data = prevData === undefined ? Object.create(null) : prevData;
    const childKey =
      (typeof result === 'string' ? result : ctx.store.keyOfEntity(result)) ||
      key;
    const selectionResult = readSelection(ctx, childKey, select, data);

    if (selectionResult !== null && typeof result === 'object') {
      for (key in result) {
        if (key !== '__typename' && result.hasOwnProperty(key)) {
          selectionResult[key] = result[key];
        }
      }
    }

    return selectionResult;
  }

  warning(
    false,
    'Invalid resolver value: The resolver at `%s` returned a scalar (number, boolean, etc)' +
      ', but the GraphQL query expects a selection set for this field.\n' +
      'If necessary, use Cache.resolve() to resolve a link or entity from the cache.',
    key
  );

  ctx.result.completeness = 'EMPTY';
  return null;
};

const resolveLink = (
  ctx: Context,
  link: Link | Link[],
  select: SelectionSet,
  prevData: void | Data | Data[]
): null | Data | Data[] => {
  if (Array.isArray(link)) {
    const newLink = new Array(link.length);
    for (let i = 0, l = link.length; i < l; i++) {
      const data = prevData !== undefined ? prevData[i] : undefined;
      newLink[i] = resolveLink(ctx, link[i], select, data);
    }

    return newLink;
  } else if (link === null) {
    return null;
  } else {
    const data = prevData === undefined ? Object.create(null) : prevData;
    return readSelection(ctx, link, select, data);
  }
};

const isDataOrKey = (x: any): x is string | Data =>
  typeof x === 'string' ||
  (typeof x === 'object' &&
    x !== null &&
    typeof (x as any).__typename === 'string');
