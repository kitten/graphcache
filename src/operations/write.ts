import warning from 'warning';
import { DocumentNode, FragmentDefinitionNode } from 'graphql';

import {
  getFieldAlias,
  getFragments,
  getMainOperation,
  getSelectionSet,
  normalizeVariables,
  getFragmentTypeName,
  getName,
  getOperationName,
  getFieldArguments,
} from '../ast';

import {
  NullArray,
  Fragments,
  Variables,
  Data,
  Link,
  Scalar,
  SelectionSet,
  OperationRequest,
} from '../types';

import {
  Store,
  addDependency,
  getCurrentDependencies,
  initStoreState,
  clearStoreState,
} from '../store';

import { forEachFieldNode } from './shared';
import { joinKeys, keyOfField } from '../helpers';

export interface WriteResult {
  dependencies: Set<string>;
}

interface Context {
  result: WriteResult;
  store: Store;
  variables: Variables;
  fragments: Fragments;
}

/** Writes a request given its response to the store */
export const write = (
  store: Store,
  request: OperationRequest,
  data: Data
): WriteResult => {
  initStoreState(0);

  const result = startWrite(store, request, data);

  clearStoreState();

  return result;
};

export const startWrite = (
  store: Store,
  request: OperationRequest,
  data: Data
) => {
  const operation = getMainOperation(request.query);
  const result: WriteResult = { dependencies: getCurrentDependencies() };

  const ctx: Context = {
    variables: normalizeVariables(operation, request.variables),
    fragments: getFragments(request.query),
    result,
    store,
  };

  const select = getSelectionSet(operation);
  const operationName = getOperationName(operation);

  if (operationName === 'Query') {
    writeSelection(ctx, operationName, select, data);
  } else {
    writeRoot(ctx, operationName, select, data);
  }

  return result;
};

export const writeOptimistic = (
  store: Store,
  request: OperationRequest,
  optimisticKey: number
): WriteResult => {
  initStoreState(optimisticKey);

  const operation = getMainOperation(request.query);
  const result: WriteResult = { dependencies: getCurrentDependencies() };

  const ctx: Context = {
    variables: normalizeVariables(operation, request.variables),
    fragments: getFragments(request.query),
    result,
    store,
  };

  const operationName = getOperationName(operation);
  if (operationName === 'Mutation') {
    const select = getSelectionSet(operation);
    forEachFieldNode(operationName, operationName, select, ctx, node => {
      if (node.selectionSet !== undefined) {
        const fieldName = getName(node);
        const resolver = ctx.store.optimisticMutations[fieldName];
        if (resolver !== undefined) {
          const fieldArgs = getFieldArguments(node, ctx.variables);
          const fieldSelect = getSelectionSet(node);
          const resolverValue = resolver(fieldArgs || {}, ctx.store, ctx);
          if (!isScalar(resolverValue)) {
            writeRootField(ctx, resolverValue, fieldSelect);
          }
        }
      }
    });
  }

  clearStoreState();
  return result;
};

export const writeFragment = (
  store: Store,
  query: DocumentNode,
  data: Data
) => {
  const fragments = getFragments(query);
  const names = Object.keys(fragments);
  const fragment = fragments[names[0]] as FragmentDefinitionNode;
  if (fragment === undefined) {
    return warning(
      false,
      'writeFragment(...) was called with an empty fragment.\n' +
        'You have to call it with at least one fragment in your GraphQL document.'
    );
  }

  const select = getSelectionSet(fragment);
  const typeName = getFragmentTypeName(fragment);
  const writeData = { ...data, __typename: typeName } as Data;

  const entityKey = store.keyOfEntity(writeData) as string;
  if (!entityKey) {
    return warning(
      false,
      "Can't generate a key for writeFragment(...) data.\n" +
        'You have to pass an `id` or `_id` field or create a custom `keys` config for `%s`.',
      typeName
    );
  }

  const ctx: Context = {
    variables: {}, // TODO: Should we support variables?
    fragments,
    result: { dependencies: getCurrentDependencies() },
    store,
  };

  writeSelection(ctx, entityKey, select, writeData);
};

const writeSelection = (
  ctx: Context,
  entityKey: string,
  select: SelectionSet,
  data: Data
) => {
  const { store, variables } = ctx;
  const isQuery = entityKey === 'Query';
  const typename = data.__typename;
  if (!isQuery) addDependency(entityKey);

  store.writeField(isQuery ? entityKey : typename, entityKey, '__typename');

  forEachFieldNode(typename, entityKey, select, ctx, node => {
    const fieldName = getName(node);
    const fieldArgs = getFieldArguments(node, variables);
    const fieldKey = joinKeys(entityKey, keyOfField(fieldName, fieldArgs));
    const fieldValue = data[getFieldAlias(node)];

    if (isQuery) addDependency(fieldKey);
    if (node.selectionSet === undefined) {
      // This is a leaf node, so we're setting the field's value directly
      store.writeRecord(fieldValue, fieldKey);
    } else if (!isScalar(fieldValue)) {
      // Process the field and write links for the child entities that have been written
      const { selections: fieldSelect } = node.selectionSet;
      const link = writeField(ctx, fieldKey, fieldSelect, fieldValue);
      store.writeLink(link, fieldKey);
      store.removeRecord(fieldKey);
    } else {
      warning(
        false,
        'Invalid value: The field at `%s` is a scalar (number, boolean, etc)' +
          ', but the GraphQL query expects a selection set for this field.\n' +
          'The value will still be cached, however this may lead to undefined behavior!',
        fieldKey
      );

      // This is a rare case for invalid entities
      store.writeRecord(fieldValue, fieldKey);
    }
  });
};

const writeField = (
  ctx: Context,
  parentFieldKey: string,
  select: SelectionSet,
  data: null | Data | NullArray<Data>
): Link => {
  if (Array.isArray(data)) {
    return data.map((item, index) => {
      // Append the current index to the parentFieldKey fallback
      const indexKey = joinKeys(parentFieldKey, `${index}`);
      // Recursively write array data
      const links = writeField(ctx, indexKey, select, item);
      // Link cannot be expressed as a recursive type
      return links as string | null;
    });
  } else if (data === null) {
    return null;
  }

  const entityKey = ctx.store.keyOfEntity(data);
  const key = entityKey !== null ? entityKey : parentFieldKey;

  warning(
    typeof data.__typename !== 'string' ||
      ctx.store.keys[data.__typename] !== undefined ||
      entityKey !== null,
    'Invalid key: The GraphQL query at the field at `%s` has a selection set, ' +
      'but no key could be generated for the data at this field.\n' +
      'You have to request `id` or `_id` fields for all selection sets or create ' +
      'a custom `keys` config for `%s`.\n' +
      'Entities without keys will be embedded directly on the parent entity. ' +
      'If this is intentional, create a `keys` config for `%s` that always returns null.',
    parentFieldKey,
    data.__typename,
    data.__typename
  );

  writeSelection(ctx, key, select, data);
  return key;
};

// This is like writeSelection but assumes no parent entity exists
const writeRoot = (
  ctx: Context,
  typename: string,
  select: SelectionSet,
  data: Data
) => {
  forEachFieldNode(typename, typename, select, ctx, node => {
    const fieldName = getName(node);
    const fieldAlias = getFieldAlias(node);
    const fieldArgs = getFieldArguments(node, ctx.variables);
    const fieldValue = data[fieldAlias];

    if (
      node.selectionSet !== undefined &&
      fieldValue !== null &&
      !isScalar(fieldValue)
    ) {
      const { selections: fieldSelect } = node.selectionSet;
      writeRootField(ctx, fieldValue, fieldSelect);
    }

    if (typename === 'Mutation' || typename === 'Subscription') {
      // We run side-effect updates after the default, normalized updates
      // so that the data is already available in-store if necessary
      const updater = ctx.store.updates[typename][fieldName];
      if (updater !== undefined) {
        updater(data, fieldArgs || {}, ctx.store, ctx);
      }
    }
  });
};

// This is like writeField but doesn't fall back to a generated key
const writeRootField = (
  ctx: Context,
  data: null | Data | NullArray<Data>,
  select: SelectionSet
) => {
  if (Array.isArray(data)) {
    return data.map(item => writeRootField(ctx, item, select));
  } else if (data === null) {
    return;
  }

  // Write entity to key that falls back to the given parentFieldKey
  const entityKey = ctx.store.keyOfEntity(data);
  if (entityKey !== null) {
    writeSelection(ctx, entityKey, select, data);
  } else {
    const typename = data.__typename;
    writeRoot(ctx, typename, select, data);
  }
};

// Without a typename field on Data or Data[] the result must be a scalar
// This effectively prevents us from writing Data into the store that
// doesn't have a __typename field
const isScalar = (x: any): x is Scalar | Scalar[] => {
  if (Array.isArray(x)) {
    return x.some(isScalar);
  }

  return (
    typeof x !== 'object' ||
    (x !== null && typeof (x as any).__typename !== 'string')
  );
};
