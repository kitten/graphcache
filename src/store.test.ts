import gql from 'graphql-tag';

import {
  Store,
  initStoreState,
  clearStoreState,
  getCurrentDependencies,
} from './store';

import { Data } from './types';
import { query } from './operations/query';
import { write, writeOptimistic } from './operations/write';

const Appointment = gql`
  query appointment($id: String) {
    appointment(id: $id) {
      __typename
      id
      info
    }
  }
`;

const Todos = gql`
  query {
    __typename
    todos {
      __typename
      id
      text
      complete
      author {
        id
        name
      }
    }
  }
`;

describe('Store with KeyingConfig', () => {
  it('generates keys from custom keying function', () => {
    const store = new Store(undefined, undefined, undefined, undefined, {
      User: () => 'me',
    });

    expect(store.keyOfEntity({ __typename: 'Any', id: '123' })).toBe('Any:123');
    expect(store.keyOfEntity({ __typename: 'Any', _id: '123' })).toBe(
      'Any:123'
    );
    expect(store.keyOfEntity({ __typename: 'Any' })).toBe(null);
    expect(store.keyOfEntity({ __typename: 'User' })).toBe('me');
  });
});

describe('Store with OptimisticMutationConfig', () => {
  let store, todosData;

  beforeEach(() => {
    store = new Store(undefined, undefined, undefined, {
      addTodo: variables => {
        return {
          ...variables,
        } as Data;
      },
    });
    todosData = {
      __typename: 'Query',
      todos: [
        {
          id: '0',
          text: 'Go to the shops',
          complete: false,
          __typename: 'Todo',
          author: { id: '0', name: 'Jovi', __typename: 'Author' },
        },
        {
          id: '1',
          text: 'Pick up the kids',
          complete: true,
          __typename: 'Todo',
          author: { id: '1', name: 'Phil', __typename: 'Author' },
        },
        {
          id: '2',
          text: 'Install urql',
          complete: false,
          __typename: 'Todo',
          author: { id: '0', name: 'Jovi', __typename: 'Author' },
        },
      ],
    };
    write(store, { query: Todos }, todosData);
    initStoreState(null);
  });

  it('Should resolve a property', () => {
    const todoResult = store.resolve({ __typename: 'Todo', id: '0' }, 'text');
    expect(todoResult).toEqual('Go to the shops');
    const authorResult = store.resolve(
      { __typename: 'Author', id: '0' },
      'name'
    );
    expect(authorResult).toBe('Jovi');
    const result = store.resolve({ id: 0, __typename: 'Todo' }, 'text');
    expect(result).toEqual('Go to the shops');
    // TODO: we have no way of asserting this to really be the case.
    const deps = getCurrentDependencies();
    expect(deps).toEqual(new Set(['Todo:0', 'Author:0']));
    clearStoreState();
  });

  it('should resolve with a key as first argument', () => {
    const authorResult = store.resolve('Author:0', 'name');
    expect(authorResult).toBe('Jovi');
    const deps = getCurrentDependencies();
    expect(deps).toEqual(new Set(['Author:0']));
    clearStoreState();
  });

  it('Should resolve a link property', () => {
    const parent = {
      id: '0',
      text: 'test',
      author: undefined,
      __typename: 'Todo',
    };
    const result = store.resolve(parent, 'author');
    expect(result).toEqual('Author:0');
    const deps = getCurrentDependencies();
    expect(deps).toEqual(new Set(['Todo:0']));
    clearStoreState();
  });

  it('should be able to invalidate data (one relation key)', () => {
    let { data } = query(store, { query: Todos });
    expect((data as any).todos).toHaveLength(3);
    expect(store.getRecord('Todo:0.text')).toBe('Go to the shops');
    initStoreState(0);
    store.invalidateQuery(Todos);
    clearStoreState();
    ({ data } = query(store, { query: Todos }));
    expect(data).toBe(null);
    expect(store.getRecord('Todo:0.text')).toBe(undefined);
  });

  it('should be able to invalidate data with arguments', () => {
    initStoreState(0);
    write(
      store,
      {
        query: Appointment,
        variables: { id: '1' },
      },
      {
        __typename: 'Query',
        appointment: {
          __typename: 'Appointment',
          id: '1',
          info: 'urql meeting',
        },
      }
    );
    clearStoreState();

    let { data } = query(store, {
      query: Appointment,
      variables: { id: '1' },
    });
    expect((data as any).appointment.info).toBe('urql meeting');
    expect(store.getRecord('Appointment:1.info')).toBe('urql meeting');
    initStoreState(0);
    store.invalidateQuery(Appointment, { id: '1' });
    clearStoreState();
    ({ data } = query(store, {
      query: Appointment,
      variables: { id: '1' },
    }));
    expect(data).toBe(null);
    expect(store.getRecord('Appointment:1.info')).toBe(undefined);
  });

  it('should be able to update a fragment', () => {
    initStoreState(0);
    store.writeFragment(
      gql`
        fragment _ on Todo {
          id
          text
          complete
        }
      `,
      {
        id: '0',
        text: 'update',
        complete: true,
      }
    );

    const deps = getCurrentDependencies();
    expect(deps).toEqual(new Set(['Todo:0']));

    const { data } = query(store, { query: Todos });

    expect(data).toEqual({
      __typename: 'Query',
      todos: [
        {
          ...todosData.todos[0],
          text: 'update',
          complete: true,
        },
        todosData.todos[1],
        todosData.todos[2],
      ],
    });
    clearStoreState();
  });

  it('should be able to update a query', () => {
    initStoreState(0);
    store.updateQuery({ query: Todos }, data => ({
      ...data,
      todos: [
        ...data.todos,
        {
          __typename: 'Todo',
          id: '4',
          text: 'Test updateQuery',
          complete: false,
          author: {
            id: '3',
            name: 'Andy',
          },
        },
      ],
    }));
    clearStoreState();

    const { data: result } = query(store, {
      query: Todos,
    });
    expect(result).toEqual({
      __typename: 'Query',
      todos: [
        ...todosData.todos,
        {
          __typename: 'Todo',
          id: '4',
          text: 'Test updateQuery',
          complete: false,
          author: {
            id: '3',
            name: 'Andy',
          },
        },
      ],
    });
    clearStoreState();
  });

  it('should be able to update a query with variables', () => {
    initStoreState(0);
    write(
      store,
      {
        query: Appointment,
        variables: { id: '1' },
      },
      {
        __typename: 'Query',
        appointment: {
          __typename: 'Appointment',
          id: '1',
          info: 'urql meeting',
        },
      }
    );
    clearStoreState();
    initStoreState(0);
    store.updateQuery({ query: Appointment, variables: { id: '1' } }, data => ({
      ...data,
      appointment: {
        ...data.appointment,
        info: 'urql meeting revisited',
      },
    }));
    clearStoreState();

    const { data: result } = query(store, {
      query: Appointment,
      variables: { id: '1' },
    });
    expect(result).toEqual({
      __typename: 'Query',
      appointment: {
        id: '1',
        info: 'urql meeting revisited',
        __typename: 'Appointment',
      },
    });
    clearStoreState();
  });

  it('should be able to optimistically mutate', () => {
    const { dependencies } = writeOptimistic(
      store,
      {
        query: gql`
          mutation {
            addTodo(
              id: "1"
              text: "I'm optimistic about this feature"
              complete: true
              __typename: "Todo"
            ) {
              id
              text
              complete
              __typename
            }
          }
        `,
      },
      1
    );
    expect(dependencies).toEqual(new Set(['Todo:1']));
    let { data } = query(store, { query: Todos });
    expect(data).toEqual({
      __typename: 'Query',
      todos: [
        todosData.todos[0],
        {
          id: '1',
          text: "I'm optimistic about this feature",
          complete: true,
          __typename: 'Todo',
          author: {
            __typename: 'Author',
            id: '1',
            name: 'Phil',
          },
        },
        todosData.todos[2],
      ],
    });

    store.clearOptimistic(1);
    ({ data } = query(store, { query: Todos }));
    expect(data).toEqual({
      __typename: 'Query',
      todos: todosData.todos,
    });
  });
});
