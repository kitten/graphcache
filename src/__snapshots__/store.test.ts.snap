// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store with storage should be able to store and rehydrate data 1`] = `
Object {
  "l|Query.todos": Array [
    "Todo:0",
  ],
  "l|Todo:0.author": "Author:0",
  "r|Author:0.__typename": "Author",
  "r|Author:0.id": "0",
  "r|Author:0.name": "Jovi",
  "r|Query.__typename": "Query",
  "r|Query.todos": undefined,
  "r|Todo:0.__typename": "Todo",
  "r|Todo:0.author": undefined,
  "r|Todo:0.complete": false,
  "r|Todo:0.id": "0",
  "r|Todo:0.text": "Go to the shops",
}
`;
